internal/compiler-bridge/src/main/scala/xsbt/ExtractAPI.scala

```diff
@@ -198,33 +202,62 @@ class ExtractAPI[GlobalType <: Global](
   // The compiler only pickles static annotations, so only include these in the API.
   // This way, the API is not sensitive to whether we compiled from source or loaded from classfile.
   // (When looking at the sources we see all annotations, but when loading from classes we only see the pickled (static) ones.)
-  private def mkAnnotations(in: Symbol, as: List[AnnotationInfo]): Array[xsbti.api.Annotation] =
-    staticAnnotations(as).toArray.map { a =>
-      xsbti.api.Annotation.of(
-        processType(in, a.atp),
-        if (a.assocs.isEmpty)
-          Array(xsbti.api.AnnotationArgument.of("", a.args.mkString("(", ",", ")"))) // what else to do with a Tree?
-        else
-          a.assocs
-            .map {
-              case (name, value) =>
-                xsbti.api.AnnotationArgument.of(name.toString, value.toString)
-            }
-            .toArray[xsbti.api.AnnotationArgument]
-      )
-    }
+  private def mkAnnotations(in: Symbol, as: List[AnnotationInfo]): Array[xsbti.api.Annotation] = {
+    if (in == NoSymbol) ExtractAPI.emptyAnnotationArray
+    else
+      staticAnnotations(as) match {
+        case Nil => ExtractAPI.emptyAnnotationArray
+        case staticAs =>
+          staticAs.map { a =>
+            xsbti.api.Annotation.of(
+              processType(in, a.atp),
+              if (a.assocs.isEmpty)
+                Array(xsbti.api.AnnotationArgument.of("", a.args.mkString("(", ",", ")"))) // what else to do with a Tree?
+              else
+                a.assocs
+                  .map {
+                    case (name, value) =>
+                      xsbti.api.AnnotationArgument.of(name.toString, value.toString)
+                  }
+                  .toArray[xsbti.api.AnnotationArgument]
+            )
+          }.toArray
+      }
+  }
 
-  private def annotations(in: Symbol, s: Symbol): Array[xsbti.api.Annotation] =
-    enteringPhase(currentRun.typerPhase) {
+  // HOT method, hand optimized to reduce allocations and needless creation of Names with calls to getterIn/setterIn
+  // on non-fields.
+  private def annotations(in: Symbol, s: Symbol): Array[xsbti.api.Annotation] = {
+    val saved = phase
+    phase = currentRun.typerPhase
```

Comment:
Is time-travelling with Scalac's API generally inefficient or impedes inlining? I see that you're inlining the body manually here.

Reply:
Does the on-the-fly compilation of the compiler interface enable the
optimiser? If not, hand inlining here is helpful.

Reply:
https://github.com/sbt/zinc/blob/v1.1.1/internal/zinc-compile-core/src/main/scala/sbt/internal/inc/AnalyzingCompiler.scala#L350:

```scala
          compiler(sourceFiles, classpath, outputDirectory, "-nowarn" :: Nil)
```

It seems like we just pass in `"-nowarn"`.

Reply:
It would be interesting to add the optimiser flags in here, WDYT?

Reply:
I think it makes sense, especially if we can observe some effect. Is there a safe flag that would work for all Scala versions, or do we have to optimize the list of optimize flag per Scala versions?

Reply:
The compiler options and bugs are version dependent. I would be inclined to leave it off, at least by default. Exposing these options as config might be useful for experimentation.